src/librustc_codegen_ssa/back/write.rs

```diff
@@ -0,0 +1,1826 @@
+    /// absolutely no optimizations (used for the metadata module).
+    pub opt_level: Option<config::OptLevel>,
+
+    /// Some(level) to optimize binary size, or None to not affect program size.
+    pub opt_size: Option<config::OptLevel>,
+
+    pub pgo_gen: Option<String>,
+    pub pgo_use: String,
+
+    // Flags indicating which outputs to produce.
+    pub emit_pre_thin_lto_bc: bool,
+    pub emit_no_opt_bc: bool,
+    pub emit_bc: bool,
+    pub emit_bc_compressed: bool,
+    pub emit_lto_bc: bool,
+    pub emit_ir: bool,
+    pub emit_asm: bool,
+    pub emit_obj: bool,
+    // Miscellaneous flags.  These are mostly copied from command-line
+    // options.
+    pub verify_llvm_ir: bool,
+    pub no_prepopulate_passes: bool,
+    pub no_builtins: bool,
+    pub time_passes: bool,
+    pub vectorize_loop: bool,
+    pub vectorize_slp: bool,
+    pub merge_functions: bool,
+    pub inline_threshold: Option<usize>,
+    // Instead of creating an object file by doing LLVM codegen, just
+    // make the object file bitcode. Provides easy compatibility with
+    // emscripten's ecc compiler, when used as the linker.
+    pub obj_is_bitcode: bool,
+    pub no_integrated_as: bool,
+    pub embed_bitcode: bool,
+    pub embed_bitcode_marker: bool,
+}
+
+impl ModuleConfig {
+    fn new(passes: Vec<String>) -> ModuleConfig {
+        ModuleConfig {
+            passes,
+            opt_level: None,
+            opt_size: None,
+
+            pgo_gen: None,
+            pgo_use: String::new(),
+
+            emit_no_opt_bc: false,
+            emit_pre_thin_lto_bc: false,
+            emit_bc: false,
+            emit_bc_compressed: false,
+            emit_lto_bc: false,
+            emit_ir: false,
+            emit_asm: false,
+            emit_obj: false,
+            obj_is_bitcode: false,
+            embed_bitcode: false,
+            embed_bitcode_marker: false,
+            no_integrated_as: false,
+
+            verify_llvm_ir: false,
+            no_prepopulate_passes: false,
+            no_builtins: false,
+            time_passes: false,
+            vectorize_loop: false,
+            vectorize_slp: false,
+            merge_functions: false,
+            inline_threshold: None
+        }
+    }
+
+    fn set_flags(&mut self, sess: &Session, no_builtins: bool) {
+        self.verify_llvm_ir = sess.verify_llvm_ir();
+        self.no_prepopulate_passes = sess.opts.cg.no_prepopulate_passes;
+        self.no_builtins = no_builtins || sess.target.target.options.no_builtins;
+        self.time_passes = sess.time_passes();
+        self.inline_threshold = sess.opts.cg.inline_threshold;
+        self.obj_is_bitcode = sess.target.target.options.obj_is_bitcode ||
+                              sess.opts.debugging_opts.cross_lang_lto.enabled();
+        let embed_bitcode = sess.target.target.options.embed_bitcode ||
+                            sess.opts.debugging_opts.embed_bitcode;
```

Comment:
Not sure what to do about all of these LLVM-specific options.

Reply:
Maybe rename `bitcode` to `ir`, to generalize? or `llvm_bc`, to make it clear it's LLVM-specific.